qt/src/Displayer.hh

```diff
@@ -65,6 +65,7 @@ public:
 
 public slots:
 	void setAngle(double angle);
+	void setScaledImage(const QImage& image, double scale = 1.0);
```

Comment:
scaled image is just the image used in the UI, not the one used when recognizing. See `Displayer::getImage`.

Reply:
It's just for testing. As far as i remember from our previous discuss about it, i should extend interface of Displayer.

Reply:
You'll probably want to do something similar to `DisplayRenderer::adjustImage`.

Reply:
Actually it depends on what operation you are performing, if they are operations which take a long time and should not be repeated everytime the image is redrawn, then you'll probably want to create a new temporary image with the operations applied, and then use that as source.

Reply:
Hah, it's quite difficult to choose, because different image operations have different duration: from milliseconds to seconds. E.g. denoising is very expensive operation.

Reply:
So I'd work on the source level directly. Do you ever need the original image again after processing? If not, just clone the original image to a temporary location (see `SourceManager::savePixmap`), and then point `Source::path` to that location.

Reply:
Hmm though you'll need some special logic for multipage documents in that case.

Reply:
Original image we need for adding "Cancel" (Ctrl+Z) for image processing algorithms. So we should store middleware images in RAM or on disk.

So isn't writing all images after every image operation huge overhead?

Reply:
>Hmm though you'll need some special logic for multipage documents in that case.

Yep. Also i need specific logic for this kind of image processing: we have one image. And on one image we have more than one page and want to split it into different pages.

Reply:
@manisandro So what should i do for setting image to displayer in right way? Also i want to use preprocessed image in OCR. Can you explain me step by step? (i looked into adjustImage - no, we can't use it)

Reply:
I'd store the processed image as a temporary image and add it as a new source.

Reply:
Where should i store this temporary image? What is the best way to add this temporary as a new source?

Reply:
See `SourceManager::savePixmap` for the temporary file logic, and then just do `SourceManager::addSource(tmpfilename)`.

Reply:
btw, what do you think about using Doxygen for source documentation? For me gIR is new project and orientation inside for me is a little bit difficult. Can i use Doxygen for documenting source code, what i write?

Reply:
To be honest I don't know if method documentation helps a lot, since the method names are pretty clear and it is pretty pointless to write i.e.
    
    /* Set the sources
     * @sources The sources
     * @return True on success, false on failure
     */
    bool setSources(const std::vector<Source*> sources);

What one needs is a general overview of how the components play together, and I don't think that  doxygen helps much there.

Reply:
You can look into image_processing sources: i use for it Doxygen.

Reply:
I'm familiar with doxygen, but again, I'm not convinced it helps much here besides blowing up the size of the source files.

Reply:
About SourceManager::addSource method: but in this case as fasr as i understand my temporary item will be added to QListWidgetItem. Is it ok?

Reply:
Another idea could be to extend the `Source` struct by adding a `QMap<int,QString> preprocessedImages`, where you store, for each page, the preprocessed image. If such a page exists, the Displayer will render that image instead of the one of the original source. You can then give the user the option to revert to the original image, in which case the entry in `preprocessedImages` is simply discarded. This approach would also make it easy to apply the algorithms to all pages of a multipage document, without polluting the sources list with tons of entries.

Reply:
I like this idea.